### PR TITLE
[4.x] Mix to Vite migration

### PR DESCRIPTION
**Pull request by esurov**
Migrated from laravel mix to vite following official migration guide: https://laravel.com/docs/10.x/vite

vite

### CODE REVIEW

File: resources/js/app.js [hunk 1]
```diff
@@ -6,8 +6,13 @@ import VueRouter from 'vue-router';
 import VueJsonPretty from 'vue-json-pretty';
 import 'vue-json-pretty/lib/styles.css';
 import moment from 'moment-timezone';
+import relatedEntries from './components/RelatedEntries.vue';
+import indexScreen from './components/IndexScreen.vue';
+import previewScreen from './components/PreviewScreen.vue';
+import alertComponent from './components/Alert.vue';
+import copyClipboard from './components/CopyClipboard.vue';
```
Comment (crynobone) [inline resources/js/app.js:13]:
```suggestion
import RelatedEntries from './components/RelatedEntries.vue';
import IndexScreen from './components/IndexScreen.vue';
import PreviewScreen from './components/PreviewScreen.vue';
import AlertComponent from './components/Alert.vue';
import CopyClipboard from './components/CopyClipboard.vue';
```

File: resources/js/app.js [hunk 2]
```diff
@@ -37,11 +43,11 @@ const router = new VueRouter({
...
-Vue.component('related-entries', require('./components/RelatedEntries.vue').default);
-Vue.component('index-screen', require('./components/IndexScreen.vue').default);
-Vue.component('preview-screen', require('./components/PreviewScreen.vue').default);
-Vue.component('alert', require('./components/Alert.vue').default);
-Vue.component('copy-clipboard', require('./components/CopyClipboard.vue').default);
+Vue.component('related-entries', relatedEntries);
+Vue.component('index-screen', indexScreen);
+Vue.component('preview-screen', previewScreen);
+Vue.component('alert', alertComponent);
+Vue.component('copy-clipboard', copyClipboard);
```
Comment (crynobone) [inline resources/js/app.js:50]:
```suggestion
Vue.component('related-entries', RelatedEntries);
Vue.component('index-screen', IndexScreen);
Vue.component('preview-screen', PreviewScreen);
Vue.component('alert', AlertComponent);
Vue.component('copy-clipboard', CopyClipboard);
```

**Review by crynobone — CHANGES_REQUESTED** [review]

File: resources/js/routes.js [hunk 3]
```diff
@@ -1,219 +1,256 @@
...
+import scheduleIndex from './screens/schedule/index.vue';
+import redisPreview from './screens/redis/preview.vue';
+import redisIndex from './screens/redis/index.vue';
+import monitoringIndex from './screens/monitoring/index.vue';
+import gatesPreview from './screens/gates/preview.vue';
+import gatesIndex from './screens/gates/index.vue';
+import viewsPreview from './screens/views/preview.vue';
+import viewsIndex from './screens/views/index.vue';
+import clientRequestsPreview from './screens/client-requests/preview.vue';
+import clientRequestsIndex from './screens/client-requests/index.vue';
```
Comment (crynobone) [inline resources/js/routes.js:36]: Should follow same `studly` case

File: resources/js/screens/logs/preview.vue [hunk 4]
```diff
@@ -1,10 +1,12 @@
 <script type="text/ecmascript-6">
     import StylesMixin from './../../mixins/entriesStyles';
+    import exceptionCodePreview from "../../components/ExceptionCodePreview.vue";
+    import stacktrace from "../../components/Stacktrace.vue";
```
Comment (crynobone) [inline resources/js/screens/logs/preview.vue:4]:
```suggestion
    import exceptionCodePreview from "@/components/ExceptionCodePreview.vue";
    import stacktrace from "@/components/Stacktrace.vue";
```

File: resources/js/screens/logs/preview.vue [hunk 5]
```diff
@@ -1,10 +1,12 @@
...
 
     export default {
         components: {
-            'code-preview': require('./../../components/ExceptionCodePreview').default,
-            'stack-trace': require('./../../components/Stacktrace').default
+            'code-preview': exceptionCodePreview,
+            'stack-trace': stacktrace
```
Comment (crynobone) [inline resources/js/screens/logs/preview.vue:9]:
```suggestion
            'code-preview': ExceptionCodePreview,
            'stack-trace': Stacktrace
```

File: resources/js/screens/jobs/preview.vue [hunk 6]
```diff
@@ -1,11 +1,13 @@
 <script type="text/ecmascript-6">
     import axios from 'axios';
     import StylesMixin from './../../mixins/entriesStyles';
+    import exceptionCodePreview from "../../components/ExceptionCodePreview.vue";
+    import stacktrace from "../../components/Stacktrace.vue";
```
Comment (crynobone) [inline resources/js/screens/jobs/preview.vue:5]:
```suggestion
    import exceptionCodePreview from "@/components/ExceptionCodePreview.vue";
    import stacktrace from "@/components/Stacktrace.vue";
```

File: resources/js/screens/jobs/preview.vue [hunk 7]
```diff
@@ -1,11 +1,13 @@
...
 
     export default {
         components: {
-            'code-preview': require('./../../components/ExceptionCodePreview').default,
-            'stack-trace': require('./../../components/Stacktrace').default
+            'code-preview': exceptionCodePreview,
+            'stack-trace': stacktrace
```
Comment (crynobone) [inline resources/js/screens/jobs/preview.vue:10]:
```suggestion
            'code-preview': ExceptionCodePreview,
            'stack-trace': Stacktrace
```

File: resources/js/screens/exceptions/preview.vue [hunk 8]
```diff
@@ -1,10 +1,12 @@
 <script type="text/ecmascript-6">
     import axios from 'axios';
+    import exceptionCodePreview from "../../components/ExceptionCodePreview.vue";
+    import stacktrace from "../../components/Stacktrace.vue";
```
Comment (crynobone) [inline resources/js/screens/exceptions/preview.vue:4]:
```suggestion
    import exceptionCodePreview from "@/components/ExceptionCodePreview.vue";
    import stacktrace from "@/components/Stacktrace.vue";
```

File: resources/js/screens/exceptions/preview.vue [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
 
     export default {
         components: {
-            'code-preview': require('./../../components/ExceptionCodePreview').default,
-            'stack-trace': require('./../../components/Stacktrace').default
+            'code-preview': exceptionCodePreview,
+            'stack-trace': stacktrace
```
Comment (crynobone) [inline resources/js/screens/exceptions/preview.vue:9]:
```suggestion
            'code-preview': ExceptionCodePreview,
            'stack-trace': Stacktrace
```

File: resources/js/routes.js (hunk 3 above)
Comment (esurov) [reply resources/js/routes.js]: Fixed

**Review by crynobone — APPROVED** [review]

**Comment by francoism90** [comment]
Hmm, I'm keep getting the following error after upgrading:
```
GET http://localhost:5173/resources/sass/app.scss net::ERR_ABORTED 404 (Not Found)
```
I've run `npm run build` and use Vite.

Horizon, etc. work fine, but Telescope is broken.

**Comment by taylorotwell** [comment]
@esurov Telescope actually doesn't even work in Laravel applications anymore?

**Comment by taylorotwell** [comment]
https://github.com/laravel/telescope/issues/1399

**Comment by wfern** [comment]
@esurov the problem is the "hot file" path...

It need to be set in place or before of the @vite directive or called somewhere before rendering.

The problem is because the default hot path is relative to the main application not telescope, so when we run `npm run dev` it gets the hot file from the main application: https://github.com/laravel/framework/blob/d0a6080db78a7e8a3316ff9c41eb86aa0866742d/src/Illuminate/Foundation/Vite.php#L170C54-L170C54

Example how to fix:
```php
// .../telescope/resources/view/layout.blade.php

// Maybe it's better to set it in the TelescopeServiceProvider but this works for test purposes 
{{ Vite::useHotFile('public/hot') }}
    
@vite('resources/sass/' . $cssFile, '/vendor/telescope/build')```

**Comment by timacdonald** [comment]
If we revisit this, we should consider: https://github.com/laravel/framework/issues/48964

**Comment by daniser** [comment]
As a person who extensively uses Vite in my own packages, I already have solutions for this problem.

### Define hot file path in template

_layout.blade.php_
```blade
{{ Vite::useHotFile('vendor/my-package/hot')
    ->useBuildDirectory('vendor/my-package/build')
    ->withEntryPoints(['resources/js/app.js', "resources/js/pages/{$page['component']}.vue"]) }}
```
The problem with this solution is, the `Vite` class is singleton, so it may affect main app entry points/build dir/hot file path etc in unpredictable ways.

### Use macro or mixin

_MyPackageServiceProvider.php_
```php
public function boot()
{
    Vite::mixin(new Support\ViteAliasMixin);
}
```
_ViteAliasMixin.php_
```php
<?php

namespace MyVendor\MyPackage\Support;

use Illuminate\Foundation\Vite;

class ViteAliasMixin
{
    protected function myPackageEntryPoint(): \Closure
    {
        return fn (array $page) => $this
            ->useHotFile('vendor/my-package/hot')
            ->useBuildDirectory('vendor/my-package/build')
            ->withEntryPoints(['resources/js/app.js', "resources/js/pages/{$page['component']}.vue"]);
    }
}
```
_layout.blade.php_
```blade
{{ Vite::myPackageEntryPoint($page) }}
```
I use this solution in my packages, but it has same flaws as the previous one.

### Implement some kind of Vite config manager and configure its instance in package's service provider
I already [had PR for this](https://github.com/laravel/framework/pull/45257), but unfortunately it was closed, so I implemented this functionality [as a package](https://github.com/ttbooking/vite-manager).
I use this in my project, but not in my packages yet, as I don't want excess dependencies. I hope maybe one day we'll revisit this PR and it'll made its way into core?

_MyPackageServiceProvider.php_
```php
public function boot()
{
    Vite::app('myapp')
        ->useHotFile('vendor/my-package/hot')
        ->useBuildDirectory('vendor/my-package/build')
        ->withEntryPoints(['resources/js/app.js']);
}
```
_layout.blade.php_
```blade
@viteApp('myapp')
```

**Pull request merged**